lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/WordStorage.java

```diff
@@ -0,0 +1,338 @@
+   *       the difference of this entry's start and the prefix's entry start. 0 for single-character
+   *       entries
+   *   <li>Optional, for non-leaf entries only:
+   *       <ul>
+   *         <li>VINT: the length of the word form data, returned from {@link #lookupWord}
+   *         <li>n * VINT: the word form data
+   *         <li>Optional, for hash-colliding entries only:
+   *             <ul>
+   *               <li>BYTE: 1 if the next collision entry has further collisions, 0 if it's the
+   *                   last of the entries with the same hash
+   *               <li>VINT: (relative) pointer to the previous entry with the same hash
+   *             </ul>
+   *       </ul>
+   * </ul>
+   */
+  private final byte[] wordData;
+
+  private WordStorage(int[] hashTable, byte[] wordData) {
+    this.hashTable = hashTable;
+    this.wordData = wordData;
+  }
+
+  IntsRef lookupWord(char[] word, int offset, int length) {
+    assert length > 0;
+
+    int hash = Math.abs(CharsRef.stringHashCode(word, offset, length) % hashTable.length);
+    int pos = hashTable[hash];
+    if (pos == 0) {
+      return null;
+    }
+
+    boolean collision = pos < 0;
+    pos = Math.abs(pos);
+
+    char lastChar = word[offset + length - 1];
+    ByteArrayDataInput in = new ByteArrayDataInput(wordData);
+    while (true) {
+      in.setPosition(pos);
+      char c = (char) in.readVInt();
+      int prevPos = pos - in.readVInt();
+      int beforeForms = in.getPosition();
+      boolean found = c == lastChar && isSameString(word, offset, length - 1, prevPos, in);
+      if (!collision && !found) {
+        return null;
+      }
+
+      in.setPosition(beforeForms);
+      int formLength = in.readVInt();
+      if (found) {
+        IntsRef forms = new IntsRef(formLength);
+        readForms(forms, in, formLength);
+        return forms;
+      } else {
+        skipVInts(in, formLength);
+      }
+
+      collision = in.readByte() == 1;
+      pos -= in.readVInt();
+    }
+  }
+
+  private static void skipVInts(ByteArrayDataInput in, int count) {
+    for (int i = 0; i < count; ) {
+      if (in.readByte() >= 0) i++;
+    }
+  }
+
+  /**
+   * @param processor is invoked for each word. Note that the passed arguments (word and form) are
+   *     reused, so they can be modified in any way, but may not be saved for later by the processor
+   */
+  void processAllWords(int maxLength, BiConsumer<CharsRef, IntsRef> processor) {
+    CharsRef chars = new CharsRef(maxLength);
+    IntsRef forms = new IntsRef();
+    ByteArrayDataInput in = new ByteArrayDataInput(wordData);
+    for (int pos : hashTable) {
+      boolean collision = pos < 0;
+      pos = Math.abs(pos);
+
+      while (pos != 0) {
+        int wordStart = maxLength - 1;
+
+        in.setPosition(pos);
+        chars.chars[wordStart] = (char) in.readVInt();
+        int prevPos = pos - in.readVInt();
+
+        int dataLength = in.readVInt();
```

Comment:
If you are willing to say "no more than 256 forms" you could make this a `.readByte()` which goes a bit faster than `.readVInt`, maybe.

Reply:
Unfortunately some crazy dictionaries have numerous homonyms, like this:

```
Alberto po:antropónimo [n-grama: Alberto Arnoldi]
Alberto po:antropónimo [n-grama: Alberto Avendaño Prieto]
Alberto po:antropónimo [n-grama: Alberto Bazzoni]
Alberto po:antropónimo [n-grama: Alberto Benito]
Alberto po:antropónimo [n-grama: Alberto Burri]
...
```

The funny thing is that we don't even store/support this custom `n-grama` thing. But Hunspell stores it, so we might, too, in future. So I'm not sure we should collapse the duplicates just for this case right now. Your opinion?